File: easy/290.js
/**
 * @param {string} pattern
 * @param {string} s
 * @return {boolean}
 */
 var wordPattern = function(pattern, s) {
    const map1 = new Map(), map2 = new Map();
    const arr = s.split(' ');
    const patt_arr = pattern.split('');
    if(arr.length !== patt_arr.length) return false;
    let f = true;
    patt_arr.every((c, i) => {
        const v1 = map1.get(c);
        const v2 = map2.get(arr[i]);
        if(v1 === undefined && v2 === undefined) {map1.set(c, arr[i]); map2.set(arr[i], c);}
        else if(v2 !== c || v1 !== arr[i]) {f = false; return false;}
        return true;
    });
    return f;
};

File: easy/290.cpp
/**
    290. Word Pattern
*/
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        unordered_map<char, string> map;
        unordered_map<string, char> map2;
        int i = 0;
        int k = 0;
        while(i < s.size()){
            int j = i;
            while(j < s.size() && s[j] != ' ')
                ++j;
            string str = s.substr(i, j - i);
            if((map[pattern[k]] != "" && str != map[pattern[k]]) || (map2[str] != 0 && map2[str] != pattern[k]))
                return false;
            else{
                map[pattern[k]] = str;
                map2[str] = pattern[k];
            }
            
            while(j < s.size() && s[j] == ' ')
                ++j;
            i = j;
            ++k;
        }
        return k < pattern.size() ? false : true;
    }
};
